BraveWallet/Crypto/Stores/SwapTokenStore.swift

```diff
@@ -0,0 +1,642 @@
+  @Published var slippageOption = SlippageGrid.Option.halfPercent {
+    didSet {
+      slippage = slippageOption.value
+    }
+  }
+  /// The custom user input slippage percentage value which will override `slippageOption` if it is not nil
+  @Published var overrideSlippage: Int? {
+    didSet {
+      if let overrideSlippage = overrideSlippage {
+        slippage = Double(overrideSlippage) / 100.0
+      } else {
+        slippage = slippageOption.value
+      }
+    }
+  }
+  
+  private let keyringController: BraveWalletKeyringController
+  private let tokenRegistry: BraveWalletERCTokenRegistry
+  private let rpcController: BraveWalletEthJsonRpcController
+  private let assetRatioController: BraveWalletAssetRatioController
+  private let swapController: BraveWalletSwapController
+  private let transactionController: BraveWalletEthTxController
+  private var accountInfo: BraveWallet.AccountInfo?
+  private var chainId: String = ""
+  private var slippage = 0.005 {
+    didSet {
+      timer?.invalidate()
+      timer = Timer.scheduledTimer(withTimeInterval: 0.25, repeats: false, block: { [weak self] _ in
+        self?.fetchPriceQuote(base: .perSellAsset)
+      })
+    }
+  }
+  private var updatingPriceQuote = false
+  private var addingUnapprovedTx = false
+  private var timer: Timer?
+  
+  enum SwapParamsBase {
+    // calculating based on sell asset amount
+    case perSellAsset
+    // calculating based on buy asset amount
+    case perBuyAsset
+  }
+  
+  enum SwapState {
+    // when there is an error occurs. associated with an error message
+    case error(String)
+    // when erc20 token's allowance is less then sell amount. associated with a spender address
+    // the future of this state would get erc20 token apporve data then create
+    // an unapproved transaction
+    case lowAllowance(String)
+    // when able to swap
+    case swap
+    // has not passed validation
+    case idle
+  }
+  
+  public init(
+    keyringController: BraveWalletKeyringController,
+    tokenRegistry: BraveWalletERCTokenRegistry,
+    rpcController: BraveWalletEthJsonRpcController,
+    assetRatioController: BraveWalletAssetRatioController,
+    swapController: BraveWalletSwapController,
+    transactionController: BraveWalletEthTxController
+  ) {
+    self.keyringController = keyringController
+    self.tokenRegistry = tokenRegistry
+    self.rpcController = rpcController
+    self.assetRatioController = assetRatioController
+    self.swapController = swapController
+    self.transactionController = transactionController
+    
+    self.keyringController.add(self)
+    self.rpcController.add(self)
+  }
+  
+  private func fetchTokenBalance(
+    for token: BraveWallet.ERCToken,
+    completion: @escaping (_ balance: BDouble?) -> Void
+  ) {
+    guard let account = accountInfo else {
+      completion(nil)
+      return
+    }
+    
+    rpcController.balance(for: token, in: account) { balance in
+      completion(BDouble(balance ?? 0))
```

Comment:
Should it really default to `0`? OR should it return `nil` since the completion block accepts `BDouble?`

Reply:
True, can return nil instead